### PR TITLE
Issue 253: Create XmlXPathRetrieve to retrieve XML data using XPath expressions

### PR DESCRIPTION
**Pull request by spetratos**
- Adding a utility to retrieve xml data that matches a given XPath
expression.

Signed-off-by: Spiros Petratos <spetratos@paypal.com>

### CODE REVIEW

File: butterfly-utilities/src/main/java/com/paypal/butterfly/utilities/operations/xml/XPathRetrieve.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+ * If the file is not a well formed XML file, an error is returned.
+ *
+ * @author spetratos
+ */
+public class XPathRetrieve extends SingleCondition<XPathRetrieve> {
+    private String xpathExpressionString;
+    private XPathExpression xpathExpression;
+    private QName returnDataType;
+
+    private static final String DESCRIPTION = "Retriveve the XML XPath query %s if it exists in XML file %s";
```
Comment (fabiocarvalho777) [inline butterfly-utilities/src/main/java/com/paypal/butterfly/utilities/operations/xml/XPathRetrieve.java]: Please, fix the typo in `Retriveve`. Also, the TU retrieves XML data, not the query itself, correct? Maybe should this be rephrased to something like: "Retrieve XML data based on the given XPath query %s and XML file %s"?

File: butterfly-utilities/src/main/java/com/paypal/butterfly/utilities/operations/xml/XPathRetrieve.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+    /**
+     * Checks if a particular xpath exists in an XML file and returns it in string format
```
Comment (fabiocarvalho777) [inline butterfly-utilities/src/main/java/com/paypal/butterfly/utilities/operations/xml/XPathRetrieve.java]: It should retrieve XML data, not the XPath query :)

**Comment by fabiocarvalho777** [comment]
Thanks for fixing the CI build @spetratos , notice that there are some other comments too. But no rush, that can be done tomorrow. Good night :-)

File: butterfly-utilities/src/test/java/com/paypal/butterfly/utilities/operations/xml/XPathRetrieveTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        assertEquals(xpathRetrieve.getDescription(), "Retriveve the XML XPath query " + XPATH_EXPRESSION +  " if it exists in XML file foo1.xml");
+    }
+
+    @Test
+    public void notFoundTest() {
+        final String XPATH_EXPRESSION = "/test";
+        XPathRetrieve xpathRetrieve = new XPathRetrieve(XPATH_EXPRESSION, XPathConstants.NODESET).relative("foo1.xml");
+        assertEquals(xpathRetrieve.getXPathExpressionString(), XPATH_EXPRESSION);
+        TUExecutionResult executionResult = xpathRetrieve.execution(transformedAppFolder, transformationContext);
+        assertEquals(executionResult.getType(), TUExecutionResult.Type.VALUE);
```
[the remark this answers is not in the document]
Comment (spetratos) [reply butterfly-utilities/src/test/java/com/paypal/butterfly/utilities/operations/xml/XPathRetrieveTest.java]: If expression is not found, XPath returns "" for String and and an empty NodeList for NODESET.
So, I did not want to set the TUExecutionResult.type to NULL or ERROR.

File: butterfly-utilities/src/test/java/com/paypal/butterfly/utilities/xml/XmlXPathRetrieveTest.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    @Test
+    public void nullTest() {
+        final String XPATH_EXPRESSION = "/foo/bar";
+        XmlXPathRetrieve xmlXPathElement = new XmlXPathRetrieve(XPATH_EXPRESSION).relative("foo.xml");
+        assertEquals(xmlXPathElement.getXPathExpressionString(), XPATH_EXPRESSION);
+        TUExecutionResult executionResult = xmlXPathElement.execution(transformedAppFolder, transformationContext);
+        assertEquals(executionResult.getType(), TUExecutionResult.Type.VALUE);
+        assertNotNull(executionResult.getValue());
```
Comment (fabiocarvalho777) [inline butterfly-utilities/src/test/java/com/paypal/butterfly/utilities/xml/XmlXPathRetrieveTest.java]: What is the value supposed to be here, a blank String? If so, please assert that.

**Pull request merged**